### PR TITLE
Implement well-defined handling of unrecoverable errors

### LINKED ISSUE #8754: iphone build - dereferenced NULL
This is purely nitpicking, when building for iOS (maybe also OSX, haven't checked) the build logs are filled with this warning repeating lots:
```
In file included from core/ustring.h:35:
core/vector.h:150:19:{150:20-150:28}: warning: binding dereferenced null pointer to reference has undefined behavior [-Wnull-dereference,2]
                        const T &aux = *((T *)0); //nullreturn

```

I'm wondering if its worth changing the build parameters to suppress this warning or rewrite the line of code. It makes it hard to find the actual errors in building :)

### PR DESCRIPTION
**Pull request by RandomShaper**
Implemented now as `__builtin_trap()` for GCC/Clang and `__debugbreak()` for MSVC, which are the recommended means of stopping a program on an unrecoverable situation.

~~Some  `_TRAP`-suffixed~~ Some  `CRASH_`-prefixed error macros added so this can be reused plus removing the need of double checks:

- `CRASH_COND`
- `CRASH_BAD_INDEX`
- `CRASH_NOW`

`GENERATE_TRAP` is implemented as well, but not intended for direct usage.

Plus transient avoidance of the flood of warnings emitted by Clang when checking `this` for `NULL`.
~~Plus removal of the pointless `while(0)` loop in some error macros.~~

Fixes #8754.

### CODE REVIEW

File: core/error_macros.h [hunk 1]
```diff
@@ -173,6 +193,18 @@ extern bool _err_error_exists;
 			_err_error_exists = false;                                                                     \
 	}
 
+/** Use this one if there is no sensible fallback, that is, the error is unrecoverable.
+ */
+
+#define ERR_FAIL_COND_TRAP(m_cond)                                                                         \
+	{                                                                                                      \
+		if (m_cond) {                                                                                      \
+			_err_print_error(FUNCTION_STR, __FILE__, __LINE__, "Condition ' " _STR(m_cond) " ' is true."); \
```
Comment (bojidar-bg) [inline core/error_macros.h]: I think you might like to add a `"Fatal: "` before `"Condition...`, since otherwise this would look the same as a `ERR_FAIL_COND` followed by a crash, while it is actually a meant-to-be crash.

Comment (RandomShaper) [inline core/error_macros.h]: Indeed.

**Comment by neikeq** [comment]
Great! Why don't we call it assert?

**Comment by RandomShaper** [comment]
Since all these ERR_FAIL_* functions are "inverse asserts" and they decided
on that naming, I just picked something that would follow.

**Comment by neikeq** [comment]
I think it would be better to use the `ERR_FAIL_` prefix for functions that fail by returning to the caller and the `ASSERT_` prefix in this case. IMO it's clearer and less verbose this way.

**Comment by RandomShaper** [comment]
I like the idea of making the name shorter.

But I insist in the fact that they have an inverted logic compared to asserts, so ASSERT_ would be misleading. What do you think about something like SEVERE_FAIL_(COND/INDEX) or even CRASH_(COND/INDEX)? And for the plain trap used when the code reaches an unexpected point, SEVERE_FAIL or CONTROLLED_CRASH.

This could open an interesting debate: is it good that Godot-style asserts check for error conditions instead of the standard practice of checking for some invariant to be met?

**Comment by RandomShaper** [comment]
Well, the index variants are indeed readable as positive conditions, but you get my point.

**Comment by RandomShaper** [comment]
Sorry for spamming. This is my proposal:
CRASH_COND(wrong_condition)
CRASH_BAD_INDEX(index, size)
CONTROLLED_CRASH()

**Comment by neikeq** [comment]
What do you mean they have inverted logic compared to asserts?

**Comment by vnen** [comment]
AIUI, asserts see a problem if the value is `false`, while Godot's `ERR_` macros see a problem if the value is `true`. I believe that's the inverted logic @RandomShaper is talking about (correct me if I'm wrong).

**Comment by RandomShaper** [comment]
Yes, that's what I mean.

**Comment by neikeq** [comment]
Oh, I see. Then, in my opinion, we should make it crash when the condition is `false` and name it `ASSERT`.

**Comment by RandomShaper** [comment]
I've updated the description of this PR (and pushed some changes).

I think it would be more convenient to postpone the idea about inverting the logic of the assertions so this one can be considered finished ASAP.

Maybe an issue to discuss that could be opened.

**Review by neikeq — APPROVED** [review]

File: core/error_macros.h [hunk 2]
```diff
@@ -173,6 +196,18 @@ extern bool _err_error_exists;
...
+/** Use this one if there is no sensible fallback, that is, the error is unrecoverable. 
+ */
+
+#define CRASH_COND(m_cond)                                                                                        \
+	{                                                                                                             \
+		if (m_cond) {                                                                                             \
+			_err_print_error(FUNCTION_STR, __FILE__, __LINE__, "FATAL: Condition ' " _STR(m_cond) " ' is true."); \
+			GENERATE_TRAP                                                                                         \
+		} else                                                                                                    \
+			_err_error_exists = false;                                                                            \
```
Comment (reduz) [inline core/error_macros.h]: This _err_error_exists is likely not needed, will just set the value unnecesarily all the time (hampering performance) and won't ad anything to when it crashes

File: core/error_macros.h [hunk 3]
```diff
@@ -115,28 +115,51 @@ extern bool _err_error_exists;
...
 /** An index has failed if m_index<0 or m_index >=m_size, the function exists.
-  * This function returns an error value, if returning Error, please select the most
-  * appropriate error condition from error_macros.h
-  */
+* This function returns an error value, if returning Error, please select the most
+* appropriate error condition from error_macros.h
+*/
 
 #define ERR_FAIL_INDEX_V(m_index, m_size, m_retval)                                                                        \
-	do {                                                                                                                   \
```
Comment (reduz) [inline core/error_macros.h]: the do { } while(0) has a reason to be
https://stackoverflow.com/questions/257418/do-while-0-what-is-it-good-for
it makes code look nicer, by adding a ; without generating warnings
so please don't remove it

File: core/object.h [hunk 4]
```diff
@@ -531,6 +531,12 @@ class Object {
 	void add_change_receptor(Object *p_receptor);
 	void remove_change_receptor(Object *p_receptor);
 
+// TODO: ensure 'this' is never NULL since it's UB, but by now, avoid warning flood
+#ifdef __clang__
```
Comment (reduz) [inline core/object.h]: this is unrelated, not sure what you are trying to do here

**Review by reduz — COMMENTED** [review]
seems ok for the most part, left some comments

File: core/error_macros.h (hunk 3 above)
Comment (karroffel) [inline core/error_macros.h]: Funny enough, it had a ; at the end already, so the warning argument is invalid 😛

Comment (RandomShaper) [inline core/error_macros.h]: I now understand the benefits of the loop construct. And we can't refactor the macro into a function because we need the preprocessor magic of converting the checked expression into a string. So I will put it back. But is this the only macro needing it?
Regarding the trailing semicolon, maybe that is due to clang-format?

File: core/object.h (hunk 4 above)
Comment (RandomShaper) [inline core/object.h]: Clang complains that 'this' can never be null in well-defined C++. And you get a flood of messages about that during compilation. I thought that instead of disabling the warning globally, it'd be better to do it just here and add a note about why and a 'to-do'. All as a transient solution until someone works out whether checking 'this' for null is actually needed or reliable, or audits the whole codebase to ensure that can never happen. The thing is that if it's UB, the compiler is allowed to remove it altogether.

File: core/error_macros.h (hunk 2 above)
Comment (RandomShaper) [inline core/error_macros.h]: Since I wasn't getting how the flag is being used exactly, I just mimicked what ERR_COND does. But if you can confirm it's pointless, I will happily remove it. :)

File: core/error_macros.h (hunk 3 above)
Comment (RandomShaper) [inline core/error_macros.h]: @karroffel is right, `ERR_FAIL_INDEX` is used everywhere with a trailing semicolon and we are getting no warnings about that from MSVC neither Clang.

File: core/error_macros.h (hunk 2 above)
Comment (RandomShaper) [inline core/error_macros.h]: Flag updates removed from the crashy macros for the next push.

**Comment by RandomShaper** [comment]
Pushed.

File: core/error_macros.h [hunk 5]
```diff
@@ -1,4 +1,4 @@
-/*************************************************************************/
+﻿/*************************************************************************/
```
Comment (akien-mga) [inline core/error_macros.h]: Did you add a BOM by any chance?

**Comment by RandomShaper** [comment]
That's probably the case.

I've switched to VS 2017 and I haven't been able yet to educate it about 
not adding them.

I'll fix it ASAP.

**Comment by RandomShaper** [comment]
BOM removed.

**Comment by RandomShaper** [comment]
`do-while(0)` constructs added back to the index error macros with a note about why.

I'm still not convinced, though. Anyway, one day someone will be able to implement them as true functions so we all are happy. :)
